spectator-web-spring/src/main/java/com/netflix/spectator/controllers/MetricsController.java

```diff
@@ -0,0 +1,229 @@
+    TagMeasurementFilter queryFilter = new TagMeasurementFilter(
+        filterMeterNameRegex, filterTagNameRegex, filterTagValueRegex);
+    MeasurementFilter filter;
+    if (all) {
+      filter = queryFilter;
+    } else {
+      filter = new ChainedMeasurementFilter(queryFilter,
+                                            getDefaultMeasurementFilter());
+    }
+
+    ApplicationRegistry response = new ApplicationRegistry();
+    response.setApplicationName(applicationName);
+    response.setApplicationVersion(applicationVersion);
+    response.setMetrics(encodeRegistry(registry, filter));
+    return response;
+  }
+
+  /**
+   * Internal API for encoding a registry that can be encoded as JSON.
+   * This is a helper function for the REST endpoint and to test against.
+   */
+  Map<String, MetricValues> encodeRegistry(
+        Registry sourceRegistry, MeasurementFilter filter) {
+    Map<String, MetricValues> metricMap = new HashMap<String, MetricValues>();
+
+    /**
+     * Spectator meters seam to group measurements. The meter name is the
+     * measurement name prefix. It seems that all the measurements within a
+     * meter instance share the same tag values as the meter.
+     * Different instances are different tag assignments, keeping the groups
+     * of measurements together.
+     *
+     * For now, we are not going to make this assumption and will just flatten
+     * out the whole measurement space. That makes fewer assumptions for now.
+     * If we make grouping assumptions we can factor out the labels into a more
+     * terse json. But the difference isnt necessarily that big and if the
+     * assumption doesnt hold, then forming the groupings will be complicated
+     * and more expensive.
+     */
+    for (Meter meter : sourceRegistry) {
+      Map<Id, List<Measurement>> collection
+          = new HashMap<Id, List<Measurement>>();
+      collectValues(collection, meter, filter);
+
+      if (!collection.isEmpty()) {
+        String kind = meterToKind(meter);
+        for (Map.Entry<Id, List<Measurement>> entry : collection.entrySet()) {
+           String entryName = entry.getKey().name();
+           MetricValues have = metricMap.get(entryName);
+           if (have == null) {
+             metricMap.put(entryName, MetricValues.make(kind, entry.getValue()));
+           } else {
+              have.addMeasurements(kind, entry.getValue());
+           }
+        }
+      }
+    }
+    return metricMap;
+  }
+
+  /**
+   * Looks up key in map, returning a specific default value if not found.
+   */
+  private static String getWithDefault(Map<String, String> map,
+                                       String key, String defaultValue) {
+    String value = map.get(key);
+    return value == null ? defaultValue : value;
+  }
+
+  /**
+   * Collect all the meter values matching the tag pattern.
+   */
+  public static List<Id> collectValues(
```

Comment:
There is already a utility function for filtering a measurement iterable:

https://github.com/Netflix/spectator/blob/master/spectator-api/src/main/java/com/netflix/spectator/api/Utils.java#L207

I'll take a look at the usage in the stackdriver PR and think about it a bit. The map bit seems a bit odd to me as in general there won't be duplicate measurement ids so each list would have a single value, but it isn't necessarily enforced right now.